src/aggregate_score.rs

```diff
@@ -43,15 +45,43 @@ impl fmt::Display for AggregateScore {
     }
 }
 
-#[allow(dead_code)]
-pub fn sample_aggregate_scores<const N_HAND: usize, R: Rng>(rng: &mut R, num_samples: u32) -> AggregateScore {
-    let mut scores = AggregateScore::default();
-    for _ in 0..num_samples {
-        scores.insert(&HandScore::from(&HandStats::from(&Hand::<N_HAND>::draw(
-            rng,
-        ))));
+fn sample<const N_HAND: usize, R: Rng>(
+    rng: &Mutex<R>,
+    aggregate: &Mutex<AggregateScore>,
+    samples: u32,
+) {
+    for _ in 0..samples {
+        let hand = Hand::<N_HAND>::draw(&mut *rng.lock().unwrap());
+        let score = HandScore::from(&HandStats::from(&hand));
+        aggregate.lock().unwrap().insert(&score);
     }
-    scores
+}
+
+#[allow(dead_code)]
+pub fn sample_aggregate_scores<const N_HAND: usize>(
+    rng: &mut StdRng,
+    samples: u32,
+    threads: u32,
+) -> AggregateScore {
+    let scores = Mutex::new(AggregateScore::default());
+    let rng = Mutex::new(rng);
+
+    let samples_per_thread = samples / threads;
+    let remainder = samples % threads;
+
+    // using scoped threads to automatically wait for everything to finish,
+    // which also allows us to use borrowed data instead of Arc
+    thread::scope(|s| {
+        for _ in 0..threads {
+            s.spawn(|| sample::<N_HAND, _>(&rng, &scores, samples_per_thread));
+        }
```

Comment:
This way of parallelizing the tasks doesn't actually work. It (should) compute the correct solution, but it will be much slower than the single threaded version. The problem is the mutex lock on scores. That mutex is locked and unlocked twice per sample, and is blocking on all of the other threads. Take a look a quick approximate benchmark:

**Single Thread**
```bash
time ./target/release/poker-stats statistics 5 10000000 1

HandScore: 
  high_card         10000000 (100.000%)
  pair              4925737  ( 49.257%)
  two_pair          476386   (  4.764%)
  three_of_a_kind   228695   (  2.287%)
  straight          35485    (  0.355%)
  flush             19834    (  0.198%)
  full_house        14548    (  0.145%)
  four_of_a_kind    2403     (  0.024%)
  straight_flush    140      (  0.001%)

real    0m0.784s
user    0m0.763s
sys     0m0.001s
```

**Four Threads**
```
time ./target/release/poker-stats statistics 5 10000000 4

HandScore: 
  high_card         10000000 (100.000%)
  pair              4932986  ( 49.330%)
  two_pair          476239   (  4.762%)
  three_of_a_kind   228787   (  2.288%)
  straight          35396    (  0.354%)
  flush             19698    (  0.197%)
  full_house        14480    (  0.145%)
  four_of_a_kind    2377     (  0.024%)
  straight_flush    139      (  0.001%)

real    0m1.498s
user    0m2.633s
sys     0m1.862s
```

I think there are two things to fix here. 

(1) Each thread should have its own RNG, rather than sharing a single RNG.  Ideally we would move that RNG into the thread when the thread is spawned, which would allow the RNG for each thread to be seeded deterministically, allowing for repeatable and deterministic unit tests. For now a quick hack would be to just grab the thread RNG inside the function that is called when the thread is spawned (and fix the repeatability thing later).

(2) Within a single thread, we need to sample many times before hitting a synchronization lock.  The "minimal" way to do this (I think), given the code that you have here, would be to create a new `AggregateScore` object inside of sample, sample all of the scores into that, and then only access the mutex on the `scores` object once, to merge the scores at the end of the call.

Reply:
If I can give an individual rng to each thread then maybe most of the comments will get solved automatically.